internal/controller/factory/statefulset.go

```diff
@@ -0,0 +1,231 @@
+	}
+
+	podEnv := []corev1.EnvVar{
+		{
+			Name: "POD_NAME",
+			ValueFrom: &corev1.EnvVarSource{
+				FieldRef: &corev1.ObjectFieldSelector{
+					FieldPath: "metadata.name",
+				},
+			},
+		},
+		{
+			Name: "POD_NAMESPACE",
+			ValueFrom: &corev1.EnvVarSource{
+				FieldRef: &corev1.ObjectFieldSelector{
+					FieldPath: "metadata.namespace",
+				},
+			},
+		},
+	}
+	podEnv = append(podEnv, cluster.Spec.PodSpec.ExtraEnv...)
+
+	statefulSet := &appsv1.StatefulSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: cluster.Namespace,
+			Name:      cluster.Name,
+		},
+		Spec: appsv1.StatefulSetSpec{
+			// initialize static fields that cannot be changed across updates.
+			Replicas:            cluster.Spec.Replicas,
+			ServiceName:         cluster.Name,
+			PodManagementPolicy: appsv1.ParallelPodManagement,
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"app.kubernetes.io/name":       "etcd",
+					"app.kubernetes.io/instance":   cluster.Name,
+					"app.kubernetes.io/managed-by": "etcd-operator",
+				},
+			},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: podMetadata,
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						{
+							Name:            "etcd",
+							Image:           "quay.io/coreos/etcd:v3.5.12",
+							ImagePullPolicy: cluster.Spec.PodSpec.ImagePullPolicy,
+							Command:         generateEtcdCommand(cluster),
+							Ports: []corev1.ContainerPort{
+								{Name: "peer", ContainerPort: 2380},
+								{Name: "client", ContainerPort: 2379},
+							},
+							Resources: cluster.Spec.PodSpec.Resources,
+							EnvFrom: []corev1.EnvFromSource{
+								{
+									ConfigMapRef: &corev1.ConfigMapEnvSource{
+										LocalObjectReference: corev1.LocalObjectReference{
+											Name: GetClusterStateConfigMapName(cluster),
+										},
+									},
+								},
+							},
+							Env: podEnv,
+							VolumeMounts: []corev1.VolumeMount{
+								{
+									Name:      "data",
+									ReadOnly:  false,
+									MountPath: "/var/run/etcd",
+								},
+							},
+							StartupProbe: &corev1.Probe{
+								ProbeHandler: corev1.ProbeHandler{
+									HTTPGet: &corev1.HTTPGetAction{
+										Path: "/readyz?serializable=false",
+										Port: intstr.FromInt32(2379),
+									},
+								},
+								InitialDelaySeconds: 1,
+								PeriodSeconds:       5,
+							},
+							LivenessProbe: &corev1.Probe{
+								ProbeHandler: corev1.ProbeHandler{
+									HTTPGet: &corev1.HTTPGetAction{
+										Path: "/livez",
+										Port: intstr.FromInt32(2379),
+									},
+								},
+								InitialDelaySeconds: 5,
+								PeriodSeconds:       5,
+							},
+							ReadinessProbe: &corev1.Probe{
+								ProbeHandler: corev1.ProbeHandler{
+									HTTPGet: &corev1.HTTPGetAction{
+										Path: "/readyz",
+										Port: intstr.FromInt32(2379),
+									},
+								},
+								InitialDelaySeconds: 5,
```

Comment:
Should we make probe settings customizable as well? 
Not for this MR. I've just thought that probe parameters could depend on infrastructure settings, volume and network latency.

Reply:
This behaviour isn't changed in this PR. I like this idea but let's make this in follow-up PR

Reply:
Follow up in #82